dhcpv4/nclient4/lease.go

```diff
@@ -0,0 +1,227 @@
+}
+
+//GetLease return the lease
+func (c *Client) GetLease() (clease DHCPv4ClientLease) {
+	clease = *c.lease
+	clease.MACAddr = c.ifaceHWAddr
+	clease.IfName = c.ifName
+	clease.ServerAddr = *c.serverAddr
+	return
+}
+
+// RequestSavingLease completes DORA handshake and store&apply the lease
+//
+// Note that modifiers will be applied *both* to Discover and Request packets.
+func (c *Client) RequestSavingLease(ctx context.Context, modifiers ...dhcpv4.Modifier) (offer, ack *dhcpv4.DHCPv4, err error) {
+	offer, err = c.DiscoverOffer(ctx, modifiers...)
+	if err != nil {
+		err = fmt.Errorf("unable to receive an offer: %w", err)
+		return
+	}
+
+	// TODO(chrisko): should this be unicast to the server?
+	request, err := dhcpv4.NewRequestFromOffer(offer, dhcpv4.PrependModifiers(modifiers,
+		dhcpv4.WithOption(dhcpv4.OptMaxMessageSize(MaxMessageSize)))...)
+	if err != nil {
+		err = fmt.Errorf("unable to create a request: %w", err)
+		return
+	}
+
+	ack, err = c.SendAndRead(ctx, c.serverAddr, request, nil)
+	if err != nil {
+		err = fmt.Errorf("got an error while processing the request: %w", err)
+		return
+	}
+	//save lease
+	c.lease = &DHCPv4ClientLease{}
+	c.lease.AssignedIP = ack.YourIPAddr
+	c.lease.AssignedIPMask = ack.SubnetMask()
+	c.lease.CreationTime = time.Now()
+	c.lease.LeaseDuration = ack.IPAddressLeaseTime(0)
+	if c.lease.LeaseDuration == 0 {
+		c.lease.LeaseDuration = defaultLeaseTime
+		c.logger.Printf("warning: server doesn't include Lease Time option or it is zero seconds, setting lease time to default %v", c.lease.LeaseDuration)
+
+	}
+	c.lease.RenewInterval = ack.IPAddressRenewalTime(0)
+	if c.lease.RenewInterval == 0 {
+		//setting default to half of lease time based on RFC2131,section 4.4.5
+		c.lease.RenewInterval = time.Duration(float64(c.lease.LeaseDuration) / 2)
+		c.logger.Printf("warning: server doesn't include Renew Time option or it is zero seconds, setting lease time to default %v", c.lease.RenewInterval)
+
+	}
+	c.lease.RebindInterval = ack.IPAddressRebindingTime(0)
+	if c.lease.RebindInterval == 0 {
+		//setting default to 0.875 of lease time based on RFC2131,section 4.4.5
+		c.lease.RebindInterval = time.Duration(float64(c.lease.LeaseDuration) * 0.875)
+		c.logger.Printf("warning: server doesn't include Renew Time option or it is zero seconds, setting lease time to default %v", c.lease.RebindInterval)
+
+	}
+	c.lease.IDOptions = dhcpv4.Options{}
+	for _, optioncode := range c.clientIDOptions {
+		v := request.Options.Get(optioncode)
+		c.lease.IDOptions.Update(dhcpv4.OptGeneric(optioncode, v))
+	}
+	c.lease.AckOptions = ack.Options
+	//update server address
+	c.serverAddr = &(net.UDPAddr{IP: ack.ServerIdentifier(), Port: 67})
+	err = c.ApplyLease(true)
+	return
+}
+
+//Release send DHCPv4 release messsage to server.
+//release is sent as unicast per RFC2131, section 4.4.4.
+//The lease need to be applied with c.ApplyLease(true) first before calling Release.
+func (c *Client) Release() error {
```

Comment:
I think if this library is to remain relatively neutral about the things I said before, I'd let the user supply the lease they want to release here:

```go
func (c *Client) Release(ack *DHCPv4) error {
  if ack == nil {
    return fmt.Errorf("no lease to release")
  }
  req, err := dhcpv4.New()
  if err != nil {
    ...
  }
  uniconn, err := net.DialUDP("udp4", &net.UDPAddr{
    IP: ack.YourIPAddr,
    Port: 68,
  }, &net.UDPAddr{
    IP: ack.ServerIdentifier(),
    Port: 67,
  })
  ...
}
```

Reply:
Shouldn't we also wait for an ACK response from the server?

Reply:
according to RFC2131, there is no response for DHCP release:
- section 3.1 the flowchart doesn't show any response to release
- section 4.3.4, no mention of response

Reply:
you're right, sorry about that. I'll modify my request to be something like:

```go
func (c *Client) Release (lease *DHCPv4ClientLease) error {
  ...
}
```

I'd still rather the client be able to release any lease, not just the one stored in it (which is why I'd rather it not be stored in the Client object either).